pkg/queryfrontend/roundtrip.go

```diff
@@ -20,60 +21,148 @@ import (
+	var (
+		queryRangeLimits, labelsLimits queryrange.Limits
+		err                            error
+	)
+	if config.QueryRangeConfig.Limits != nil {
+		queryRangeLimits, err = validation.NewOverrides(*config.QueryRangeConfig.Limits, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "initialize query range limits")
+		}
+	}
 
-	queriesCount := promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
-		Name: "thanos_query_frontend_queries_total",
-		Help: "Total queries passing through query frontend",
-	}, []string{"op"})
-	queriesCount.WithLabelValues(labelQuery)
-	queriesCount.WithLabelValues(labelQueryRange)
+	if config.LabelsConfig.Limits != nil {
+		labelsLimits, err = validation.NewOverrides(*config.LabelsConfig.Limits, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "initialize labels limits")
+		}
+	}
 
-	limits, err := validation.NewOverrides(*config.CortexLimits, nil)
+	queryRangeCodec := NewThanosQueryRangeCodec(config.QueryRangeConfig.PartialResponseStrategy)
+	labelsCodec := NewThanosLabelsCodec(config.LabelsConfig.PartialResponseStrategy, config.DefaultTimeRange)
+
+	queryRangeTripperware, err := newQueryRangeTripperware(config.QueryRangeConfig, queryRangeLimits, queryRangeCodec,
+		prometheus.WrapRegistererWith(prometheus.Labels{"tripperware": "query_range"}, reg), logger)
 	if err != nil {
-		return nil, errors.Wrap(err, "initialize limits")
+		return nil, err
+	}
+
+	labelsTripperware := newLabelsTripperware(config.LabelsConfig, labelsLimits, labelsCodec,
+		prometheus.WrapRegistererWith(prometheus.Labels{"tripperware": "labels"}, reg), logger)
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return newRoundTripper(next, queryRangeTripperware(next), labelsTripperware(next), reg)
+	}, nil
+}
+
+type roundTripper struct {
+	next, queryRange, metadata http.RoundTripper
+
+	queriesCount *prometheus.CounterVec
+}
+
+func newRoundTripper(next, queryRange, metadata http.RoundTripper, reg prometheus.Registerer) roundTripper {
+	r := roundTripper{
+		next:       next,
+		queryRange: queryRange,
+		metadata:   metadata,
+		queriesCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_query_frontend_queries_total",
+			Help: "Total queries passing through query frontend",
+		}, []string{"op"}),
+	}
+
+	r.queriesCount.WithLabelValues(instantQueryOp)
+	r.queriesCount.WithLabelValues(rangeQueryOp)
+	r.queriesCount.WithLabelValues(labelNamesOp)
+	r.queriesCount.WithLabelValues(labelValuesOp)
+	r.queriesCount.WithLabelValues(seriesOp)
+	return r
+}
+
+func (r roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	switch op := getOperation(req); op {
+	case instantQueryOp:
+		r.queriesCount.WithLabelValues(instantQueryOp).Inc()
+	case rangeQueryOp:
+		r.queriesCount.WithLabelValues(rangeQueryOp).Inc()
+		return r.queryRange.RoundTrip(req)
+	case labelNamesOp, labelValuesOp, seriesOp:
+		r.queriesCount.WithLabelValues(op).Inc()
+		return r.metadata.RoundTrip(req)
+	default:
```

Comment:
How about adding an `Unknown` label for this case?

Reply:
I think not needed. Since query frontend is just a proxy for downstream querier, so all other requests are unknown. Like the requests for frontend resources... Then the unknown requests would be a lot.